solr/jetty/src/java/org/apache/solr/jetty/QueuedThreadPool.java

```diff
@@ -0,0 +1,1178 @@
+    {
+        for (Thread thread : _threads)
+        {
+            if (thread.getId() == id)
+            {
+                StringBuilder buf = new StringBuilder();
+                buf.append(thread.getId()).append(" ").append(thread.getName()).append(" ");
+                buf.append(thread.getState()).append(":").append(System.lineSeparator());
+                for (StackTraceElement element : thread.getStackTrace())
+                {
+                    buf.append("  at ").append(element.toString()).append(System.lineSeparator());
+                }
+                return buf.toString();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public String toString()
+    {
+        long count = _counts.get();
+        int threads = Math.max(0, AtomicBiInteger.getHi(count));
+        int idle = Math.max(0, AtomicBiInteger.getLo(count));
+        int queue = getQueueSize();
+
+        return String.format("%s[%s]@%x{%s,%d<=%d<=%d,i=%d,r=%d,q=%d}[%s]",
+            getClass().getSimpleName(),
+            _name,
+            hashCode(),
+            getState(),
+            getMinThreads(),
+            threads,
+            getMaxThreads(),
+            idle,
+            getReservedThreads(),
+            queue,
+            _tryExecutor);
+    }
+
+    private class Runner implements Runnable
+    {
+        private Runnable idleJobPoll(long idleTimeout) throws InterruptedException
+        {
+            if (idleTimeout <= 0)
+                return _jobs.take();
+            return _jobs.poll(idleTimeout, TimeUnit.MILLISECONDS);
+        }
+
+        @Override
+        public void run()
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Runner started for {}", QueuedThreadPool.this);
+
+            boolean idle = true;
+            boolean pruneIdle = false;
+            try
+            {
+                Runnable job = null;
+                pruneIdle = shrinkManager.onIdle();
+                while (true)
+                {
+                    // If we had a job,
+                    if (job != null)
+                    {
+                        idle = true;
+                        // signal that we are idle again
+                        if (!addCounts(0, 1))
+                            break;
+                    }
+                    // else check we are still running
+                    else if (_counts.getHi() == Integer.MIN_VALUE)
+                    {
+                        break;
+                    }
+
+                    try
+                    {
+                        // Look for an immediately available job
+                        job = _jobs.poll();
+                        if (job == null)
+                        {
+                            // No job immediately available maybe we should shrink?
+                            long idleTimeout = getIdleTimeout();
+                            if (idleTimeout > 0 && getThreads() > _minThreads)
```

Comment:
The other usecase is here

1. what if all threads checks this condition same time and goes for pollIdleShrink() method
2. Now, is it possible pollIdleShrink returns `true` for all the threads? if getShrinkInterval() is very small and each thread able to set `lastShrink` ?

Reply:
True, good catch! I [dug into this a bit on the upstream PR](https://github.com/eclipse/jetty.project/pull/9498#discussion_r1149362479) (thanks for reporting there!), and even writing a test specifically designed to trigger this case (thread count falling below minThreads), I was only ever able to get it to drop 1 thread below, and the `ensureThreads()` method in the `finally` block of the run loop brings the thread count back in line with configured `minThreads`.

I'm inclined to go with the upstream, which appears to be leaning towards not worrying about this.